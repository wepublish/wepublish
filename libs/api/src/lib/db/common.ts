import {PageInfo} from '@wepublish/utils/api'

export interface ConnectionResult<T> {
  nodes: T[]
  pageInfo: PageInfo
  totalCount: number
}

export const DefaultSessionTTL = 1000 * 60 * 60 * 24 * 7 // 1w
export const DefaultBcryptHashCostFactor = 11

export enum DateFilterComparison {
  GreaterThan = 'gt',
  GreaterThanOrEqual = 'gte',
  Equal = 'eq',
  LowerThan = 'lt',
  LowerThanOrEqual = 'lte'
}
export interface DateFilter {
  date: Date | null
  comparison: DateFilterComparison
}
