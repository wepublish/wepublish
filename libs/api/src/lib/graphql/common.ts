import { DateFilterComparison, SortOrder } from '@wepublish/utils/api';
import {
  GraphQLBoolean,
  GraphQLEnumType,
  GraphQLInputObjectType,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLString,
} from 'graphql';
import { GraphQLDateTime } from 'graphql-scalars';

export const GraphQLSortOrder = new GraphQLEnumType({
  name: 'SortOrder',
  values: {
    [SortOrder.Ascending]: { value: SortOrder.Ascending },
    [SortOrder.Descending]: { value: SortOrder.Descending },
  },
});

export const GraphQLPageInfo = new GraphQLObjectType({
  name: 'PageInfo',
  fields: {
    startCursor: { type: GraphQLString },
    endCursor: { type: GraphQLString },
    hasNextPage: { type: new GraphQLNonNull(GraphQLBoolean) },
    hasPreviousPage: { type: new GraphQLNonNull(GraphQLBoolean) },
  },
});

export const GraphQLUnidirectionalPageInfo = new GraphQLObjectType({
  name: 'UnidirectionalPageInfo',
  fields: {
    endCursor: { type: GraphQLString },
    hasNextPage: { type: new GraphQLNonNull(GraphQLBoolean) },
  },
});

export const GraphQLMetadataProperty = new GraphQLObjectType({
  name: 'Properties',
  fields: {
    key: { type: new GraphQLNonNull(GraphQLString) },
    value: { type: new GraphQLNonNull(GraphQLString) },
    public: { type: new GraphQLNonNull(GraphQLBoolean) },
  },
});

export const GraphQLMetadataPropertyPublic = new GraphQLObjectType({
  name: 'PublicProperties',
  fields: {
    key: { type: new GraphQLNonNull(GraphQLString) },
    value: { type: new GraphQLNonNull(GraphQLString) },
  },
});

export const GraphQLMetadataPropertyInput = new GraphQLInputObjectType({
  name: 'PropertiesInput',
  fields: {
    key: { type: new GraphQLNonNull(GraphQLString) },
    value: { type: new GraphQLNonNull(GraphQLString) },
    public: { type: new GraphQLNonNull(GraphQLBoolean) },
  },
});

export const GraphQLMetadataPropertyPublicInput = new GraphQLInputObjectType({
  name: 'PublicPropertiesInput',
  fields: {
    key: { type: new GraphQLNonNull(GraphQLString) },
    value: { type: new GraphQLNonNull(GraphQLString) },
  },
});

export const GraphQLDateFilterComparison = new GraphQLEnumType({
  name: 'DateFilterComparison',
  values: {
    [DateFilterComparison.GreaterThan]: {
      value: DateFilterComparison.GreaterThan,
    },
    [DateFilterComparison.GreaterThanOrEqual]: {
      value: DateFilterComparison.GreaterThanOrEqual,
    },
    [DateFilterComparison.Equal]: { value: DateFilterComparison.Equal },
    [DateFilterComparison.LowerThan]: { value: DateFilterComparison.LowerThan },
    [DateFilterComparison.LowerThanOrEqual]: {
      value: DateFilterComparison.LowerThanOrEqual,
    },
  },
});

export const GraphQLDateFilter = new GraphQLInputObjectType({
  name: 'DateFilter',
  fields: {
    date: { type: GraphQLDateTime, defaultValue: null },
    comparison: { type: new GraphQLNonNull(GraphQLDateFilterComparison) },
  },
});
