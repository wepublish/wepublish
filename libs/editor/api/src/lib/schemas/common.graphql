

fragment MetadataProperty on Properties {
    key
    value
    public
}
