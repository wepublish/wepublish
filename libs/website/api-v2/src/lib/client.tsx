import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
  ApolloLink,
  NormalizedCacheObject
} from '@apollo/client'
import {ComponentType, memo, useMemo} from 'react'
import {BatchHttpLink} from '@apollo/client/link/batch-http'

export const getV2ApiClient = (
  apiUrl: string,
  links: ApolloLink[],
  cache?: NormalizedCacheObject
) => {
  const httpLink = new BatchHttpLink({
    uri: `${apiUrl}/v2`,
    batchMax: 5,
    batchInterval: 20
  })
  const link = [...links, httpLink].reduce(
    (links: ApolloLink | undefined, link) => (links ? links.concat(link) : link),
    undefined
  )

  return new ApolloClient({
    link,
    cache: new InMemoryCache().restore(cache ?? {}),
    ssrMode: typeof window === 'undefined'
  })
}

export const useV2ApiClient = (
  apiUrl: string,
  links: ApolloLink[] = [],
  cache?: NormalizedCacheObject
) => useMemo(() => getV2ApiClient(apiUrl, links, cache), [apiUrl, links, cache])

export const createWithV2ApiClient =
  (apiUrl: string, links: ApolloLink[] = [], cache?: NormalizedCacheObject) =>
  <
    // eslint-disable-next-line @typescript-eslint/ban-types
    P extends object
  >(
    ControlledComponent: ComponentType<P>
  ) =>
    memo<P>(props => {
      const client = useV2ApiClient(apiUrl, links, cache)

      return (
        <ApolloProvider client={client}>
          <ControlledComponent {...props} />
        </ApolloProvider>
      )
    })
