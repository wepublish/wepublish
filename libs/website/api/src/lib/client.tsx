import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
  ApolloLink,
  NormalizedCacheObject
} from '@apollo/client'
import {BatchHttpLink} from '@apollo/client/link/batch-http'

import {ComponentType, memo, useMemo} from 'react'

export const getV1ApiClient = (
  apiUrl: string,
  links: ApolloLink[],
  cache?: NormalizedCacheObject
) => {
  const httpLink = new BatchHttpLink({uri: `${apiUrl}/v1`, batchMax: 5, batchInterval: 20})
  const link = [...links, httpLink].reduce(
    (links: ApolloLink | undefined, link) => (links ? links.concat(link) : link),
    undefined
  )

  return new ApolloClient({
    link,
    cache: new InMemoryCache().restore(cache ?? {}),
    ssrMode: typeof window === 'undefined'
  })
}

export const useV1ApiClient = (
  apiUrl: string,
  links: ApolloLink[] = [],
  cache?: NormalizedCacheObject
) => useMemo(() => getV1ApiClient(apiUrl, links, cache), [apiUrl, links, cache])

export const createWithV1ApiClient =
  (apiUrl: string, links: ApolloLink[] = [], cache?: NormalizedCacheObject) =>
  <
    // eslint-disable-next-line @typescript-eslint/ban-types
    P extends object
  >(
    ControlledComponent: ComponentType<P>
  ) =>
    memo<P>(props => {
      const client = useV1ApiClient(apiUrl, links, cache)

      return (
        <ApolloProvider client={client}>
          <ControlledComponent {...props} />
        </ApolloProvider>
      )
    })
