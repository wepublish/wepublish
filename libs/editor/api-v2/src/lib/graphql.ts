// THIS FILE IS AUTOGENERATED
import {Node} from 'slate'
import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
const defaultOptions = {} as const;
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type Query = {
  __typename?: 'Query';
  versionInformation: VersionInformation;
};

export type VersionInformation = {
  __typename?: 'VersionInformation';
  version: Scalars['String'];
};

export type VersionInformationQueryVariables = Exact<{ [key: string]: never; }>;


export type VersionInformationQuery = { __typename?: 'Query', versionInformation: { __typename?: 'VersionInformation', version: string } };


export const VersionInformationDocument = gql`
    query VersionInformation {
  versionInformation {
    version
  }
}
    `;

/**
 * __useVersionInformationQuery__
 *
 * To run a query within a React component, call `useVersionInformationQuery` and pass it any options that fit your needs.
 * When your component renders, `useVersionInformationQuery` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useVersionInformationQuery({
 *   variables: {
 *   },
 * });
 */
export function useVersionInformationQuery(baseOptions?: Apollo.QueryHookOptions<VersionInformationQuery, VersionInformationQueryVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useQuery<VersionInformationQuery, VersionInformationQueryVariables>(VersionInformationDocument, options);
      }
export function useVersionInformationLazyQuery(baseOptions?: Apollo.LazyQueryHookOptions<VersionInformationQuery, VersionInformationQueryVariables>) {
          const options = {...defaultOptions, ...baseOptions}
          return Apollo.useLazyQuery<VersionInformationQuery, VersionInformationQueryVariables>(VersionInformationDocument, options);
        }
export type VersionInformationQueryHookResult = ReturnType<typeof useVersionInformationQuery>;
export type VersionInformationLazyQueryHookResult = ReturnType<typeof useVersionInformationLazyQuery>;
export type VersionInformationQueryResult = Apollo.QueryResult<VersionInformationQuery, VersionInformationQueryVariables>;