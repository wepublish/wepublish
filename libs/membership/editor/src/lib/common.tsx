import {ApolloClient, ApolloError, NormalizedCacheObject} from '@apollo/client'
import {TFunction} from 'i18next'
import {Message, toaster} from 'rsuite'

const showErrors = (error: ApolloError): void => {
  toaster.push(
    <Message type="error" showIcon closable duration={3000}>
      {error.message}
    </Message>
  )
}

const showSavedToast = (t: TFunction): void => {
  toaster.push(
    <Message type="success" showIcon closable duration={3000}>
      {t('subscriptionFlow.savedChange').toString()}
    </Message>
  )
}

/**
 * Default options for the GraphQL client. Displays errors and a completion message.
 * @param client the graphql client to make the request with
 * @param t the translation instance
 * @returns QueryHookOptions for the GraphQL client
 */
export const DEFAULT_MUTATION_OPTIONS = (
  client: ApolloClient<NormalizedCacheObject>,
  t: TFunction
) => {
  return {
    ...DEFAULT_QUERY_OPTIONS(client),
    onCompleted: () => showSavedToast(t)
  }
}

/**
 * Default options for the GraphQL client. Displays errors.
 * @param client the graphql client to make the request with
 * @returns QueryHookOptions for the GraphQL client
 */
export const DEFAULT_QUERY_OPTIONS = (client: ApolloClient<NormalizedCacheObject>) => {
  return {
    client,
    onError: showErrors
  }
}
