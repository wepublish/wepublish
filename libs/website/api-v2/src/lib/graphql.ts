// THIS FILE IS AUTOGENERATED
import {Node} from 'slate'
import {gql} from '@apollo/client'
import * as Apollo from '@apollo/client'
export type Maybe<T> = T | null
export type InputMaybe<T> = Maybe<T>
export type Exact<T extends {[key: string]: unknown}> = {[K in keyof T]: T[K]}
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & {[SubKey in K]?: Maybe<T[SubKey]>}
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & {[SubKey in K]: Maybe<T[SubKey]>}
const defaultOptions = {} as const
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string
  String: string
  Boolean: boolean
  Int: number
  Float: number
  /** A date-time string at UTC, such as 2019-12-03T09:54:33Z, compliant with the date-time format. */
  DateTime: string
}

export type DashboardInvoice = {
  __typename?: 'DashboardInvoice'
  amount: Scalars['Int']
  dueAt: Scalars['DateTime']
  memberPlan?: Maybe<Scalars['String']>
  paidAt?: Maybe<Scalars['DateTime']>
}

export type DashboardSubscription = {
  __typename?: 'DashboardSubscription'
  deactivationDate?: Maybe<Scalars['DateTime']>
  endsAt?: Maybe<Scalars['DateTime']>
  memberPlan: Scalars['String']
  monthlyAmount: Scalars['Int']
  paymentPeriodicity: PaymentPeriodicity
  reasonForDeactivation?: Maybe<SubscriptionDeactivationReason>
  renewsAt?: Maybe<Scalars['DateTime']>
  startsAt: Scalars['DateTime']
}

export enum PaymentPeriodicity {
  Biannual = 'biannual',
  Monthly = 'monthly',
  Quarterly = 'quarterly',
  Yearly = 'yearly'
}

export type Query = {
  __typename?: 'Query'
  /**
   * Returns all active subscribers.
   * Includes subscribers with a cancelled but not run out subscription.
   */
  activeSubscribers: Array<DashboardSubscription>
  /**
   * Returns the expected revenue for the time period given.
   * Excludes cancelled or manually set as paid invoices.
   */
  expectedRevenue: Array<DashboardInvoice>
  /**
   * Returns all new deactivations in a given timeframe.
   * This considers the time the deactivation was made, not when the subscription runs out.
   */
  newDeactivations: Array<DashboardSubscription>
  /**
   * Returns all new subscribers in a given timeframe.
   * Includes already deactivated ones.
   */
  newSubscribers: Array<DashboardSubscription>
  /** Returns all renewing subscribers in a given timeframe. */
  renewingSubscribers: Array<DashboardSubscription>
  /**
   * Returns the revenue generated for the time period given.
   * Only includes paid invoices that have not been manually paid.
   */
  revenue: Array<DashboardInvoice>
  versionInformation: VersionInformation
}

export type QueryExpectedRevenueArgs = {
  end?: InputMaybe<Scalars['DateTime']>
  start: Scalars['DateTime']
}

export type QueryNewDeactivationsArgs = {
  end?: InputMaybe<Scalars['DateTime']>
  start: Scalars['DateTime']
}

export type QueryNewSubscribersArgs = {
  end?: InputMaybe<Scalars['DateTime']>
  start: Scalars['DateTime']
}

export type QueryRenewingSubscribersArgs = {
  end?: InputMaybe<Scalars['DateTime']>
  start: Scalars['DateTime']
}

export type QueryRevenueArgs = {
  end?: InputMaybe<Scalars['DateTime']>
  start: Scalars['DateTime']
}

export enum SubscriptionDeactivationReason {
  InvoiceNotPaid = 'invoiceNotPaid',
  None = 'none',
  UserSelfDeactivated = 'userSelfDeactivated'
}

export type VersionInformation = {
  __typename?: 'VersionInformation'
  version: Scalars['String']
}

export type RevenueQueryVariables = Exact<{
  start: Scalars['DateTime']
  end?: InputMaybe<Scalars['DateTime']>
}>

export type RevenueQuery = {
  __typename?: 'Query'
  revenue: Array<{
    __typename?: 'DashboardInvoice'
    amount: number
    paidAt?: string | null
    memberPlan?: string | null
  }>
}

export type VersionInformationQueryVariables = Exact<{[key: string]: never}>

export type VersionInformationQuery = {
  __typename?: 'Query'
  versionInformation: {__typename?: 'VersionInformation'; version: string}
}

export const RevenueDocument = gql`
  query Revenue($start: DateTime!, $end: DateTime) {
    revenue(start: $start, end: $end) {
      amount
      paidAt
      memberPlan
    }
  }
`

/**
 * __useRevenueQuery__
 *
 * To run a query within a React component, call `useRevenueQuery` and pass it any options that fit your needs.
 * When your component renders, `useRevenueQuery` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useRevenueQuery({
 *   variables: {
 *      start: // value for 'start'
 *      end: // value for 'end'
 *   },
 * });
 */
export function useRevenueQuery(
  baseOptions: Apollo.QueryHookOptions<RevenueQuery, RevenueQueryVariables>
) {
  const options = {...defaultOptions, ...baseOptions}
  return Apollo.useQuery<RevenueQuery, RevenueQueryVariables>(RevenueDocument, options)
}
export function useRevenueLazyQuery(
  baseOptions?: Apollo.LazyQueryHookOptions<RevenueQuery, RevenueQueryVariables>
) {
  const options = {...defaultOptions, ...baseOptions}
  return Apollo.useLazyQuery<RevenueQuery, RevenueQueryVariables>(RevenueDocument, options)
}
export type RevenueQueryHookResult = ReturnType<typeof useRevenueQuery>
export type RevenueLazyQueryHookResult = ReturnType<typeof useRevenueLazyQuery>
export type RevenueQueryResult = Apollo.QueryResult<RevenueQuery, RevenueQueryVariables>
export const VersionInformationDocument = gql`
  query VersionInformation {
    versionInformation {
      version
    }
  }
`

/**
 * __useVersionInformationQuery__
 *
 * To run a query within a React component, call `useVersionInformationQuery` and pass it any options that fit your needs.
 * When your component renders, `useVersionInformationQuery` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useVersionInformationQuery({
 *   variables: {
 *   },
 * });
 */
export function useVersionInformationQuery(
  baseOptions?: Apollo.QueryHookOptions<VersionInformationQuery, VersionInformationQueryVariables>
) {
  const options = {...defaultOptions, ...baseOptions}
  return Apollo.useQuery<VersionInformationQuery, VersionInformationQueryVariables>(
    VersionInformationDocument,
    options
  )
}
export function useVersionInformationLazyQuery(
  baseOptions?: Apollo.LazyQueryHookOptions<
    VersionInformationQuery,
    VersionInformationQueryVariables
  >
) {
  const options = {...defaultOptions, ...baseOptions}
  return Apollo.useLazyQuery<VersionInformationQuery, VersionInformationQueryVariables>(
    VersionInformationDocument,
    options
  )
}
export type VersionInformationQueryHookResult = ReturnType<typeof useVersionInformationQuery>
export type VersionInformationLazyQueryHookResult = ReturnType<
  typeof useVersionInformationLazyQuery
>
export type VersionInformationQueryResult = Apollo.QueryResult<
  VersionInformationQuery,
  VersionInformationQueryVariables
>
