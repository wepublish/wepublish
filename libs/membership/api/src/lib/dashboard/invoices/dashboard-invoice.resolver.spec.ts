import {Test, TestingModule} from '@nestjs/testing'
import {INestApplication, Module} from '@nestjs/common'
import request from 'supertest'
import {GraphQLModule} from '@nestjs/graphql'
import {ApolloDriverConfig, ApolloDriver} from '@nestjs/apollo'
import {Invoice, PrismaClient} from '@prisma/client'
import {PrismaModule} from '@wepublish/nest-modules'
import {DashboardInvoiceResolver} from './dashboard-invoice.resolver'
import {DashboardInvoiceService} from './dashboard-invoice.service'

@Module({
  imports: [
    GraphQLModule.forRoot<ApolloDriverConfig>({
      driver: ApolloDriver,
      autoSchemaFile: true,
      path: '/',
      cache: 'bounded'
    }),
    PrismaModule
  ],
  providers: [DashboardInvoiceResolver, DashboardInvoiceService]
})
export class AppModule {}

const revenueQuery = `
  query Dashboard($end:DateTime!, $start:DateTime!) {
    revenue(start:$start, end:$end) {
      amount
      dueAt
      memberPlan
      paidAt
    }
  }
`

const expectedRevenueQuery = `
  query Dashboard($end:DateTime!, $start:DateTime!) {
    expectedRevenue(start:$start, end:$end) {
      amount
      dueAt
      memberPlan
      paidAt
    }
  }
`

describe('DashboardInvoiceResolver', () => {
  let invoicesToDelete: Invoice[] = []
  let app: INestApplication
  let prisma: PrismaClient

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [AppModule]
    }).compile()

    prisma = module.get<PrismaClient>(PrismaClient)
    app = module.createNestApplication()
    await app.init()
  })

  beforeEach(async () => {
    invoicesToDelete = []
  })

  afterEach(async () => {
    const ids = invoicesToDelete.map(invoice => invoice.id)

    await prisma.invoice.deleteMany({
      where: {
        id: {
          in: ids
        }
      }
    })
  })

  afterAll(async () => {
    await app.close()
  })

  test('revenue', async () => {
    const mockData = [
      {
        dueAt: new Date('2023-01-01 12:00:00'),
        paidAt: new Date('2023-01-01 12:00:00'),
        mail: 'foo@wepublish.ch',
        items: {
          createMany: {
            data: [
              {
                amount: 50,
                name: '',
                quantity: 1
              },
              {
                amount: 500,
                name: '',
                quantity: 2
              }
            ]
          }
        }
      },
      {
        dueAt: new Date('2023-01-01 12:00:00'),
        paidAt: new Date('2023-01-02 12:00:00'),
        mail: 'foo@wepublish.ch',
        items: {
          createMany: {
            data: [
              {
                amount: 50,
                name: '',
                quantity: 1
              }
            ]
          }
        }
      },
      {
        dueAt: new Date('2023-01-01 12:00:00'),
        paidAt: new Date('2023-02-01 12:00:00'),
        mail: 'foo@wepublish.ch',
        items: {
          createMany: {
            data: [
              {
                amount: 50,
                name: '',
                quantity: 1
              }
            ]
          }
        }
      },
      {
        dueAt: new Date('2023-01-01 12:00:00'),
        paidAt: new Date('2023-01-01 12:00:00'),
        mail: 'foo@wepublish.ch',
        manuallySetAsPaidByUserId: '123',
        items: {
          createMany: {
            data: [
              {
                amount: 50,
                name: '',
                quantity: 1
              }
            ]
          }
        }
      },
      {
        dueAt: new Date('2023-01-01 12:00:00'),
        mail: 'foo@wepublish.ch',
        items: {
          createMany: {
            data: [
              {
                amount: 50,
                name: '',
                quantity: 1
              }
            ]
          }
        }
      }
    ]

    for (const data of mockData) {
      invoicesToDelete.push(await prisma.invoice.create({data}))
    }

    await request(app.getHttpServer())
      .post('')
      .send({
        query: revenueQuery,
        variables: {
          start: new Date('2023-01-01').toISOString(),
          end: new Date('2023-02-01').toISOString()
        }
      })
      .expect(200)
      .expect(res => {
        expect(res.body.data.revenue).toMatchSnapshot()
      })
  })

  test('expectedRevenue', async () => {
    const mockData = [
      {
        dueAt: new Date('2023-01-01 12:00:00'),
        paidAt: new Date('2023-01-01 12:00:00'),
        mail: 'foo@wepublish.ch',
        items: {
          createMany: {
            data: [
              {
                amount: 50,
                name: '',
                quantity: 1
              },
              {
                amount: 500,
                name: '',
                quantity: 2
              }
            ]
          }
        }
      },
      {
        dueAt: new Date('2023-01-02 12:00:00'),
        mail: 'foo@wepublish.ch',
        items: {
          createMany: {
            data: [
              {
                amount: 50,
                name: '',
                quantity: 1
              }
            ]
          }
        }
      },
      {
        dueAt: new Date('2023-02-01 12:00:00'),
        mail: 'foo@wepublish.ch',
        items: {
          createMany: {
            data: [
              {
                amount: 50,
                name: '',
                quantity: 1
              }
            ]
          }
        }
      },
      {
        dueAt: new Date('2023-01-01 12:00:00'),
        mail: 'foo@wepublish.ch',
        manuallySetAsPaidByUserId: '123',
        items: {
          createMany: {
            data: [
              {
                amount: 50,
                name: '',
                quantity: 1
              }
            ]
          }
        }
      },
      {
        dueAt: new Date('2023-01-01 12:00:00'),
        mail: 'foo@wepublish.ch',
        canceledAt: new Date('2023-01-01 11:00:00'),
        items: {
          createMany: {
            data: [
              {
                amount: 50,
                name: '',
                quantity: 1
              }
            ]
          }
        }
      }
    ]

    for (const data of mockData) {
      invoicesToDelete.push(await prisma.invoice.create({data}))
    }

    await request(app.getHttpServer())
      .post('')
      .send({
        query: expectedRevenueQuery,
        variables: {
          start: new Date('2023-01-01').toISOString(),
          end: new Date('2023-02-01').toISOString()
        }
      })
      .expect(200)
      .expect(res => {
        expect(res.body.data.expectedRevenue).toMatchSnapshot()
      })
  })
})
