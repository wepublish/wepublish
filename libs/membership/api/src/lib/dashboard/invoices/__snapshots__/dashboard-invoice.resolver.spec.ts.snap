// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardInvoiceResolver expectedRevenue 1`] = `
[
  {
    "amount": 50,
    "dueAt": "2023-01-02T12:00:00.000Z",
    "memberPlan": null,
    "paidAt": null,
  },
  {
    "amount": 1050,
    "dueAt": "2023-01-01T12:00:00.000Z",
    "memberPlan": null,
    "paidAt": "2023-01-01T12:00:00.000Z",
  },
]
`;

exports[`DashboardInvoiceResolver revenue 1`] = `
[
  {
    "amount": 50,
    "dueAt": "2023-01-01T12:00:00.000Z",
    "memberPlan": null,
    "paidAt": "2023-01-02T12:00:00.000Z",
  },
  {
    "amount": 1050,
    "dueAt": "2023-01-01T12:00:00.000Z",
    "memberPlan": null,
    "paidAt": "2023-01-01T12:00:00.000Z",
  },
]
`;
